socs/agents/meinberg_syncbox/agent.py

```diff
@@ -0,0 +1,492 @@
+import argparse
+import os
+import time
+
+import txaio
+from autobahn.twisted.util import sleep as dsleep
+from ocs import ocs_agent, site_config
+from ocs.ocs_twisted import TimeoutLock
+from twisted.internet.defer import inlineCallbacks
+
+from socs.snmp import SNMPTwister
+
+# For logging
+txaio.use_twisted()
+
+
+def _extract_oid_field_and_value(get_result):
+    """Extract field names and OID values from SNMP GET results.
+
+    The ObjectType objects returned from pysnmp interactions contain the
+    info we want to use for field names, specifically the OID and associated
+    integer for uniquely identifying duplicate OIDs, as well as the value of the
+    OID, which we want to save.
+
+    Here we use the prettyPrint() method to get the OID name, requiring
+    some string manipulation. We also just grab the hidden ._value
+    directly, as this seemed the cleanest way to get an actual value of a
+    normal type. Without doing this we get a pysnmp defined Integer32 or
+    DisplayString, which were awkward to handle, particularly the
+    DisplayString.
+
+    Parameters
+    ----------
+    get_result : pysnmp.smi.rfc1902.ObjectType
+        Result from a pysnmp GET command.
+
+    Returns
+    -------
+    field_name : str
+        Field name for an OID, i.e. 'outletStatus_1'
+    oid_value : int or str
+        Associated value for the OID. Returns None if not an int or str
+    oid_description : str
+        String description of the OID value.
+    """
+    # OID from SNMP GET
+    oid = get_result[0].prettyPrint()
+    # Makes something like 'IBOOTPDU-MIB::outletStatus.1'
+    # look like 'outletStatus_1'
+    field_name = oid.split("::")[1].replace('.', '_')
+
+    # Grab OID value, mostly these are integers
+    oid_value = get_result[1]._value
+    oid_description = get_result[1].prettyPrint()
+
+    # Decode string values
+    if isinstance(oid_value, bytes):
+        oid_value = oid_value.decode("utf-8")
+
+    # I don't expect any other types at the moment, but just in case.
+    if not isinstance(oid_value, (int, bytes, str)):
+        oid_value = None
+
+    return field_name, oid_value, oid_description
+
+
+def _build_message(get_result, time):
+    """Build the message for publication on an OCS Feed.
+
+    Parameters
+    ----------
+    get_result : pysnmp.smi.rfc1902.ObjectType
+        Result from a pysnmp GET command.
+    time : float
+        Timestamp for when the SNMP GET was issued.
+
+    Returns
+    -------
+    message : dict
+        OCS Feed formatted message for publishing
+    """
+    message = {
+        'block_name': 'syncbox',
+        'timestamp': time,
+        'data': {}
+    }
+
+    for item in get_result:
+        field_name, oid_value, oid_description = _extract_oid_field_and_value(item)
+
+        if oid_value is None:
+            oid_value = 'No SNMP response'
```

Comment:
But the usual value can be integer or float for some fields, right?  In that case you can't mix strings in (influx hates that ... and probably aggregator won't like it either).